pkg/kv/kvserver/store.go

```diff
@@ -411,8 +411,277 @@ func (rs *storeReplicaVisitor) EstimatedCount() int {
+
+Ranges
+
+Each Replica is part of a Range, i.e. corresponds to what other systems would
+call a shard. A Range is a consensus group backed by Raft, i.e. each Replica
+is a state machine backed by a replicated log of commands. In CockroachDB,
+Ranges own a contiguous chunk of addressable keyspace, where the word
+"addressable" is a fine-print that interested readers can learn about in
+keys.Addr; in short the Replica data is logically contiguous, but not
+contiguous when viewed as Engine kv pairs.
+
+Considerable complexity is incurred by the features of dynamic re-sharding
+and relocation, i.e. range splits, range merges, and range relocation (also
+called replication changes, or, in the case of lateral movement,
+rebalancing). Each of these interact heavily with the Range a consensus group
+(of which each Replica is a member). All of these intricacies are described
+at a high level in this comment.
+
+RangeDescriptor
+
+A roachpb.RangeDescriptor is the configuration of a Range. It is an
+MVCC-backed key-value pair (where the key is derived from the StartKey via
+keys.RangeDescriptorKey, which in particular resides on the Range itself)
+that is accessible to the transactional KV API much like any other, but is
+for internal use only (and in particular plays no role at the SQL layer).
+
+Splits, Merges, and Rebalances are all carried out as distributed
+transactions. They are all complex in their own right but they share the
+basic approach of transactionally acting on the RangeDescriptor. Each of
+these operations at some point will
+
+- start a transaction
+
+- update the RangeDescriptor (for example, to reflect a split, or a change
+to the Replicas comprising the members of the Range)
+
+- update the meta ranges (which form a search index used for request routing)
+
+- commit with a roachpb.InternalCommitTrigger.
+
+In particular, note that the RangeDescriptor is the first write issued in the
+transaction, and so the transaction record will be created on the affected
+range. This allows us to establish a helpful invariant:
+
+INVARIANT: an intent on keys.RangeDescriptorKey is resolved atomically with
+the (application of the) roachpb.EndTxnRequest committing the transaction.
+
+To understand this invariant, it's best to try to make do without it and fail.
+Assume that a descriptor change committed, but that the intent will only get
+resolved asynchronously.
+
+- since the transaction committed, its update to the meta ranges becomes
+visible and the cluster will start routing requests according to the new
+descriptor. However, the Replicas may yet have to learn that the new
+descriptor is committed. This is not a correctness problem, but it is
+undesirable. The invariant avoids this, since the first intent to be resolved
+will be the RangeDescriptor on the leaseholder (during EndTxn application),
+and this will causally precede resolution of auxiliary intents since those
+will only be resolved after the leaseholder has applied the EndTxn.
```

Comment:
This isn't entirely clear to me. Is the only benefit that we don't have to wait in line for intent resolution? I suppose the resolution also involves a write, and thus another consensus round. But it's essentially to avoid any additional latency in discovering the new descriptor, right?

Reply:
It's a weak argument. As my TODO below mentions, I thought there were stronger reasons than the ones presented here, though having to attach logic to intent resolution (the next bullet point) is probably enough to do it this way.